File: win/KesselCli/PlayWindow.cs
using System.Drawing;
using System.Drawing.Imaging;
using System.Drawing.Drawing2D;
using System.Runtime.InteropServices;
using System.Windows.Forms;
using uniffi.kessel_core;

namespace KesselCli;

// Windows analogue of the macOS play window (swift/Sources/KesselCli/PlayWindow.swift).
// `kessel --play <file.ux|file.asm>` loads a ROM into a standalone VmPlayer (no LLM
// needed), opens a WinForms window, and renders the 128x128 framebuffer scaled up
// with nearest-neighbour at 60 Hz, mapping the keyboard to the gamepad.
internal static class PlayWindow
{
    /// Load `romPath` and run the game window. Returns a process exit code.
    public static int Run(string romPath)
    {
        string source;
        try
        {
            source = File.ReadAllText(romPath);
        }
        catch (Exception e)
        {
            Console.Error.WriteLine($"Cannot read '{romPath}': {e.Message}");
            return 1;
        }

        var player = new VmPlayer();
        var error = player.Load(source, romPath);
        if (!string.IsNullOrEmpty(error))
        {
            Console.Error.WriteLine($"Failed to load '{romPath}':\n{error}");
            return 1;
        }

        // WinForms requires an STA thread; the console entry point is MTA, so run
        // the UI on a dedicated STA thread and wait for it.
        int exitCode = 0;
        var ui = new Thread(() =>
        {
            try
            {
                Application.EnableVisualStyles();
                Application.SetCompatibleTextRenderingDefault(false);
                Console.Error.WriteLine(
                    "[play] arrows move, Z/X = A/B, Enter/Space = Start/Select. Close the window to quit.");
                Application.Run(new PlayForm(player, Path.GetFileName(romPath)));
            }
            catch (Exception e)
            {
                Console.Error.WriteLine($"Play window error: {e.Message}");
                exitCode = 1;
            }
        });
        ui.SetApartmentState(ApartmentState.STA);
        ui.Start();
        ui.Join();
        return exitCode;
    }
}

/// The game window: a 60 Hz timer ticks the VM and blits the framebuffer.
internal sealed class PlayForm : Form
{
    private readonly VmPlayer _player;
    private readonly int _dim;
    private readonly System.Windows.Forms.Timer _timer;
    private readonly Bitmap _frame;
    private readonly byte[] _bgra;
    private byte _pressed;

    public PlayForm(VmPlayer player, string title)
    {
        _player = player;
        _dim = (int)player.ScreenDim();
        _frame = new Bitmap(_dim, _dim, PixelFormat.Format32bppArgb);
        _bgra = new byte[_dim * _dim * 4];

        const int scale = 5;
        Text = $"Kessel — {title}";
        ClientSize = new Size(_dim * scale, _dim * scale);
        FormBorderStyle = FormBorderStyle.FixedSingle;
        MaximizeBox = false;
        StartPosition = FormStartPosition.CenterScreen;
        DoubleBuffered = true;
        KeyPreview = true;

        _timer = new System.Windows.Forms.Timer { Interval = 16 }; // ~60 Hz
        _timer.Tick += (_, _) => Step();
        _timer.Start();
    }

    private void Step()
    {
        _player.Tick(_pressed);
        var rgba = _player.FramebufferRgba();
        if (rgba is { Length: > 0 })
        {
            WriteFramebuffer(rgba);
            Invalidate();
        }
    }

    // Our buffer is RGBA; GDI+ Format32bppArgb is BGRA in memory — swap R/B.
    private void WriteFramebuffer(byte[] rgba)
    {
        int n = _dim * _dim;
        for (int i = 0; i < n; i++)
        {
            int j = i * 4;
            _bgra[j + 0] = rgba[j + 2]; // B
            _bgra[j + 1] = rgba[j + 1]; // G
            _bgra[j + 2] = rgba[j + 0]; // R
            _bgra[j + 3] = rgba[j + 3]; // A
        }
        var rect = new Rectangle(0, 0, _dim, _dim);
        var data = _frame.LockBits(rect, ImageLockMode.WriteOnly, PixelFormat.Format32bppArgb);
        Marshal.Copy(_bgra, 0, data.Scan0, _bgra.Length);
        _frame.UnlockBits(data);
    }

    protected override void OnPaint(PaintEventArgs e)
    {
        e.Graphics.InterpolationMode = InterpolationMode.NearestNeighbor;
        e.Graphics.PixelOffsetMode = PixelOffsetMode.Half;
        e.Graphics.DrawImage(_frame, ClientRectangle);
    }

    private static byte? Bit(Keys k) => k switch
    {
        Keys.Left => (byte)0x01,
        Keys.Right => (byte)0x02,
        Keys.Up => (byte)0x04,
        Keys.Down => (byte)0x08,
        Keys.Z => (byte)0x10,
        Keys.X => (byte)0x20,
        Keys.Enter => (byte)0x40,
        Keys.Space => (byte)0x80,
        _ => null,
    };

    protected override void OnKeyDown(KeyEventArgs e)
    {
        var b = Bit(e.KeyCode);
        if (b.HasValue) { _pressed |= b.Value; e.Handled = true; }
        else base.OnKeyDown(e);
    }

    protected override void OnKeyUp(KeyEventArgs e)
    {
        var b = Bit(e.KeyCode);
        if (b.HasValue) { _pressed &= (byte)~b.Value; e.Handled = true; }
        else base.OnKeyUp(e);
    }

    protected override void OnFormClosed(FormClosedEventArgs e)
    {
        _timer.Stop();
        _timer.Dispose();
        _frame.Dispose();
        base.OnFormClosed(e);
    }
}

File: swift/Sources/KesselCli/PlayWindow.swift
import AppKit
import Foundation
import AgentBridge

// Phase 2 of the fantasy-console VM: a native, human-playable window. `kessel
// --play <file.ux|file.asm>` loads a ROM into a standalone `VmPlayer` (no LLM
// needed) and renders its 128x128 framebuffer scaled up, ticking at 60 Hz and
// mapping the keyboard to the gamepad. The same ROMs the model authors headless
// are playable here.

/// Gamepad button bits — must match `device.rs`.
private enum Btn {
    static let left: UInt8 = 0x01
    static let right: UInt8 = 0x02
    static let up: UInt8 = 0x04
    static let down: UInt8 = 0x08
    static let a: UInt8 = 0x10
    static let b: UInt8 = 0x20
    static let start: UInt8 = 0x40
    static let select: UInt8 = 0x80
}

/// A view that draws the current RGBA framebuffer with nearest-neighbour scaling
/// and tracks pressed keys as a gamepad bitfield.
final class PixelView: NSView {
    var image: CGImage?
    private(set) var pressed: UInt8 = 0

    // Top-left origin so the framebuffer's first row renders at the top.
    override var isFlipped: Bool { true }
    override var acceptsFirstResponder: Bool { true }

    override func draw(_ dirtyRect: NSRect) {
        guard let ctx = NSGraphicsContext.current?.cgContext else { return }
        ctx.setFillColor(NSColor.black.cgColor)
        ctx.fill(bounds)
        if let img = image {
            ctx.interpolationQuality = .none
            ctx.draw(img, in: bounds)
        }
    }

    private func bit(for keyCode: UInt16) -> UInt8? {
        switch keyCode {
        case 123: return Btn.left
        case 124: return Btn.right
        case 126: return Btn.up
        case 125: return Btn.down
        case 6: return Btn.a       // z
        case 7: return Btn.b       // x
        case 36: return Btn.start  // return
        case 49: return Btn.select // space
        default: return nil
        }
    }

    override func keyDown(with event: NSEvent) {
        if event.isARepeat { return }
        if let b = bit(for: event.keyCode) { pressed |= b } else { super.keyDown(with: event) }
    }

    override func keyUp(with event: NSEvent) {
        if let b = bit(for: event.keyCode) { pressed &= ~b } else { super.keyUp(with: event) }
    }

    /// Clear held keys when focus is lost, so nothing "sticks".
    override func resignFirstResponder() -> Bool {
        pressed = 0
        return true
    }
}

/// Drives one `VmPlayer` at 60 Hz into a `PixelView`.
final class PlayController {
    private let player: VmPlayer
    private let view: PixelView
    private let dim: Int
    private var timer: Timer?

    init(player: VmPlayer, view: PixelView) {
        self.player = player
        self.view = view
        self.dim = Int(player.screenDim())
    }

    func start() {
        let t = Timer(timeInterval: 1.0 / 60.0, repeats: true) { [weak self] _ in
            self?.tick()
        }
        RunLoop.main.add(t, forMode: .common) // keep ticking during window resize/menus
        timer = t
    }

    func stop() {
        timer?.invalidate()
        timer = nil
    }

    private func tick() {
        player.tick(buttons: view.pressed)
        if let data = player.framebufferRgba() {
            view.image = Self.makeImage(data: data, dim: dim)
            view.needsDisplay = true
        }
    }

    private static func makeImage(data: Data, dim: Int) -> CGImage? {
        guard data.count == dim * dim * 4, let provider = CGDataProvider(data: data as CFData) else {
            return nil
        }
        return CGImage(
            width: dim,
            height: dim,
            bitsPerComponent: 8,
            bitsPerPixel: 32,
            bytesPerRow: dim * 4,
            space: CGColorSpaceCreateDeviceRGB(),
            bitmapInfo: CGBitmapInfo(rawValue: CGImageAlphaInfo.premultipliedLast.rawValue),
            provider: provider,
            decode: nil,
            shouldInterpolate: false,
            intent: .defaultIntent
        )
    }
}

/// Quit the app when the game window closes.
final class PlayAppDelegate: NSObject, NSApplicationDelegate {
    func applicationShouldTerminateAfterLastWindowClosed(_ sender: NSApplication) -> Bool { true }
}

// Retain the AppKit objects for the process lifetime (main-thread only).
nonisolated(unsafe) private var playObjects: [Any] = []

/// Load `romPath` into a standalone player and run the game window. Blocks in the
/// AppKit run loop and exits the process when the window closes; never returns.
@MainActor
func runPlayMode(romPath: String) -> Never {
    let source: String
    do {
        source = try String(contentsOfFile: romPath, encoding: .utf8)
    } catch {
        FileHandle.standardError.write(Data("Cannot read '\(romPath)': \(error)\n".utf8))
        exit(1)
    }

    let player = VmPlayer()
    let error = player.load(source: source, path: romPath)
    if !error.isEmpty {
        FileHandle.standardError.write(Data("Failed to load '\(romPath)':\n\(error)\n".utf8))
        exit(1)
    }

    let app = NSApplication.shared
    app.setActivationPolicy(.regular)

    let dim = Int(player.screenDim())
    let scale = 5
    let side = CGFloat(dim * scale)
    let frame = NSRect(x: 0, y: 0, width: side, height: side)

    let view = PixelView(frame: frame)
    let window = NSWindow(
        contentRect: frame,
        styleMask: [.titled, .closable, .miniaturizable],
        backing: .buffered,
        defer: false
    )
    window.title = "Kessel — \(URL(fileURLWithPath: romPath).lastPathComponent)"
    window.contentView = view
    window.center()
    window.isReleasedWhenClosed = false
    window.makeKeyAndOrderFront(nil)
    window.makeFirstResponder(view)

    let controller = PlayController(player: player, view: view)
    let delegate = PlayAppDelegate()
    app.delegate = delegate
    playObjects = [controller, delegate, window]

    print("Playing \(romPath) — arrows move, Z/X = A/B, Return/Space = Start/Select. Close the window to quit.")
    controller.start()
    app.activate(ignoringOtherApps: true)
    app.run()
    exit(0)
}
